src/test-helpers.tsx

```diff
@@ -907,3 +908,28 @@ export const dispatchComment = ({
 } = {}) => {
   return dispatchComments({ comments: [comment], ...params });
 };
+
+export const createFakeApiPage = ({
+  next = null,
+  previous = null,
+  page_size = 25,
+  results = [],
+  count = results.length,
+  page_count = count > page_size ? Math.ceil(count / page_size) : 1,
+}: // eslint-disable-next-line @typescript-eslint/no-explicit-any
+Partial<PaginatedResponse<any>> = {}): PaginatedResponse<any> => {
```

Comment:
Instead of `any`, could you pass a generic type so that in `createFakeCommentsResponse` you can call `createFakeApiPage<ExternalComment>(...)`?

Reply:
Because of the JSX `extends` workaround needed, I'm not sure what to use as the base. I feel like it might work with the JSX workaround. Any ideas?

First attempt:

<details>
<summary>Patch</summary>

```diff
diff --git a/src/api/index.tsx b/src/api/index.tsx
index 19facf0..4399d43 100644
--- a/src/api/index.tsx
+++ b/src/api/index.tsx
@@ -352,7 +352,7 @@ export const createOrUpdateComment = async ({
   });
 };
 
-type GetCommentsResponse = PaginatedResponse<ExternalComment[]>;
+export type GetCommentsResponse = PaginatedResponse<ExternalComment[]>;
 
 export const getComments = async ({
   _callApi = callApi,
diff --git a/src/test-helpers.tsx b/src/test-helpers.tsx
index ce37a66..8926bb0 100644
--- a/src/test-helpers.tsx
+++ b/src/test-helpers.tsx
@@ -9,7 +9,7 @@ import { Store } from 'redux';
 import log from 'loglevel';
 import { createAction } from 'typesafe-actions';
 
-import { PaginatedResponse } from './api';
+import { GetCommentsResponse, PaginatedResponse } from './api';
 import configureStore, { ThunkActionCreator } from './configureStore';
 import { ApplicationState } from './reducers';
 import {
@@ -909,15 +909,14 @@ export const dispatchComment = ({
   return dispatchComments({ comments: [comment], ...params });
 };
 
-export const createFakeApiPage = ({
+export const createFakeApiPage = <PaginatedResponseType extends {}>({
   next = null,
   previous = null,
   page_size = 25,
   results = [],
   count = results.length,
   page_count = count > page_size ? Math.ceil(count / page_size) : 1,
-}: // eslint-disable-next-line @typescript-eslint/no-explicit-any
-Partial<PaginatedResponse<any>> = {}): PaginatedResponse<any> => {
+}: Partial<PaginatedResponseType> = {}): PaginatedResponseType => {
   return {
     count,
     next,
@@ -931,5 +930,5 @@ Partial<PaginatedResponse<any>> = {}): PaginatedResponse<any> => {
 export const createFakeCommentsResponse = (
   results = [createFakeExternalComment()],
 ) => {
-  return createFakeApiPage({ results });
+  return createFakeApiPage<GetCommentsResponse>({ results });
 };
```

</details>

<details>
<summary>Error</summary>

```
src/test-helpers.tsx:913:3 - error TS2339: Property 'next' does not exist on type 'Partial<PaginatedResponseType>'.

913   next = null,
      ~~~~

src/test-helpers.tsx:914:3 - error TS2339: Property 'previous' does not exist on type 'Partial<PaginatedResponseType>'.

914   previous = null,
      ~~~~~~~~

src/test-helpers.tsx:915:3 - error TS2339: Property 'page_size' does not exist on type 'Partial<PaginatedResponseType>'.

915   page_size = 25,
      ~~~~~~~~~

src/test-helpers.tsx:916:3 - error TS2339: Property 'results' does not exist on type 'Partial<PaginatedResponseType>'.

916   results = [],
      ~~~~~~~

src/test-helpers.tsx:917:3 - error TS2339: Property 'count' does not exist on type 'Partial<PaginatedResponseType>'.

917   count = results.length,
      ~~~~~

src/test-helpers.tsx:918:3 - error TS2339: Property 'page_count' does not exist on type 'Partial<PaginatedResponseType>'.

918   page_count = count > page_size ? Math.ceil(count / page_size) : 1,
      ~~~~~~~~~~

src/test-helpers.tsx:920:3 - error TS2322: Type '{ count: any; next: any; page_count: any; page_size: any; previous: any; results: any; }' is not assignable to type 'PaginatedResponseType'.
  '{ count: any; next: any; page_count: any; page_size: any; previous: any; results: any; }' is assignable to the constraint of type 'PaginatedResponseType', but 'PaginatedResponseType' could be instantiated with a different subtype of constraint '{}'.

920   return {
      ~~~~~~~~
921     count,
    ~~~~~~~~~~
...
926     results,
    ~~~~~~~~~~~~
927   };
    ~~~~

[10:45:12 AM] Found 7 errors. Watching for file changes.
```

</details>

Second attempt:

<details>
<summary>Patch</summary>

```diff
diff --git a/src/api/index.tsx b/src/api/index.tsx
index 19facf0..4bd4086 100644
--- a/src/api/index.tsx
+++ b/src/api/index.tsx
@@ -352,7 +352,8 @@ export const createOrUpdateComment = async ({
   });
 };
 
-type GetCommentsResponse = PaginatedResponse<ExternalComment[]>;
+export type GetCommentsResultsType = ExternalComment[];
+type GetCommentsResponse = PaginatedResponse<GetCommentsResultsType>;
 
 export const getComments = async ({
   _callApi = callApi,
diff --git a/src/test-helpers.tsx b/src/test-helpers.tsx
index ce37a66..58010e9 100644
--- a/src/test-helpers.tsx
+++ b/src/test-helpers.tsx
@@ -9,7 +9,7 @@ import { Store } from 'redux';
 import log from 'loglevel';
 import { createAction } from 'typesafe-actions';
 
-import { PaginatedResponse } from './api';
+import { GetCommentsResultsType, PaginatedResponse } from './api';
 import configureStore, { ThunkActionCreator } from './configureStore';
 import { ApplicationState } from './reducers';
 import {
@@ -909,15 +909,14 @@ export const dispatchComment = ({
   return dispatchComments({ comments: [comment], ...params });
 };
 
-export const createFakeApiPage = ({
+export const createFakeApiPage = <ResultsType extends any[]>({
   next = null,
   previous = null,
   page_size = 25,
   results = [],
   count = results.length,
   page_count = count > page_size ? Math.ceil(count / page_size) : 1,
-}: // eslint-disable-next-line @typescript-eslint/no-explicit-any
-Partial<PaginatedResponse<any>> = {}): PaginatedResponse<any> => {
+}: Partial<PaginatedResponse<ResultsType>> = {}): PaginatedResponse<ResultsType> => {
   return {
     count,
     next,
@@ -931,5 +930,5 @@ Partial<PaginatedResponse<any>> = {}): PaginatedResponse<any> => {
 export const createFakeCommentsResponse = (
   results = [createFakeExternalComment()],
 ) => {
-  return createFakeApiPage({ results });
+  return createFakeApiPage<GetCommentsResultsType>({ results });
 };
```

</details>

<details>
<summary>Errors</summary>

```
src/test-helpers.tsx:916:3 - error TS2322: Type 'never[]' is not assignable to type 'ResultsType'.
  'never[]' is assignable to the constraint of type 'ResultsType', but 'ResultsType' could be instantiated with a different subtype of constraint 'any[]'.

916   results = [],
      ~~~~~~~

[10:51:35 AM] Found 1 error. Watching for file changes.
```

</details>

Reply:
Aha, I figured out a way with a type assertion 🤷‍♂  

See https://github.com/mozilla/addons-code-manager/pull/1092/commits/745d7ce78169ed935732843cc1f3a5141577f9a5

Anyway, it adds essential type safety which I was testing with this:

```diff
diff --git a/src/test-helpers.tsx b/src/test-helpers.tsx
index 9299903..169b9a5 100644
--- a/src/test-helpers.tsx
+++ b/src/test-helpers.tsx
@@ -934,3 +934,5 @@ export const createFakeCommentsResponse = (
 ) => {
   return createFakeApiPage<GetCommentsResponse>({ results });
 };
+
+console.log(createFakeCommentsResponse().results[0].nope);

```

Thanks for the reminder on this. I meant to revisit it 😬